src/util/converter/factory.js

```diff
@@ -0,0 +1,146 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2022 (original work) Open Assessment Technologies SA ;
+ */
+
+import _ from 'lodash';
+
+/**
+ * Defines a converter processor runtime.
+ * @callback converterProcessorRuntime
+ * @param {string} text - The text to convert.
+ * @param {object} [config] - An optional config object that may contain processor specific configuration.
+ * @returns {string} - Returns the converted text.
+ */
+
+/**
+ * Defines a converter processor.
+ * @typedef {object} converterProcessor
+ * @property {string} name - The name of the converter processor. It needs to be unique within the collection.
+ * @property {converterProcessorRuntime} convert - The processor runtime that will actually convert the text.
+ */
+
+/**
+ * Checks a converter processor is valid.
+ * @param {converterProcessor} processor - The converter processor to validate.
+ * @returns {boolean} - Returns `true` if the converter processor is valid ; returns `false` otherwise.
+ */
+function validateProcessor(processor) {
+    if (!_.isPlainObject(processor)) {
+        return false;
+    }
+
+    if ('string' !== typeof processor.name || !processor.name) {
+        return false;
+    }
+
+    return 'function' === typeof processor.convert;
+}
+
+/**
+ * Creates a text converter.
+ * @param {converterProcessor[]} builtinProcessors - A list of built-in converter processors.
+ * @param {object} [builtinConfig] - An optional default config object that may contain processor specific configuration.
+ *                                   It will be forwarded to each call to the converter.
+ * @returns {converter} - Returns the text converter, ready for use.
+ * @export 'util/converter/factory'
+ */
+export default function converterFactory(builtinProcessors = [], builtinConfig = {}) {
+    let processors = [];
+
+    // find the index of the named processor
+    const findProcessor = name => processors.findIndex(processor => processor.name === name);
+
+    /**
+     * @typedef {object} converter
+     */
+    const converter = {
+        /**
+         * Converts a text with respect to the registered converter processors.
+         * @param {string} text - The text to convert.
+         * @param {object} [config] - An optional config object that may contain processor specific configuration.
+         *                            It will be merged with the possible builtin config.
+         * @returns {string} - Returns the converted text.
+         */
+        convert(text, config = {}) {
+            const localConfig = Object.assign({}, builtinConfig, config);
```

Comment:
As far as I understand there is no guarantee for config to be rewrited, say we have two different processors with `separator` config property (like in the test cases). I think we can scope configs by converter name, what do you think?

Or is it intended behaviour?

Reply:
Yes, it was for simplifying the design. Nothing prevents scoping the config per processor. And nothing prevents sharing config as well. Then, I wanted to give the possibility to overwrite the config when calling the converter for more flexibility. As an example, we could use this component for handling the localization of numbers (one converter for replacing commas with dots, and one another for the reciprocal).